Plugins/NativeEngine/Source/ShaderCompilerTraversers.cpp

```diff
@@ -495,89 +495,189 @@ namespace Babylon::ShaderCompilerTraversers
-                {
-                    "a_position",
-                    "a_normal",
-                    "a_tangent",
-                    "a_bitangent",
-                    "a_color0",
-                    "a_color1",
-                    "a_color2",
-                    "a_color3",
-                    "a_indices",
-                    "a_weight",
-                    "a_texcoord0",
-                    "a_texcoord1",
-                    "a_texcoord2",
-                    "a_texcoord3",
-                    "a_texcoord4",
-                    "a_texcoord5",
-                    "a_texcoord6",
-                    "a_texcoord7",
-                };
-            BX_STATIC_ASSERT(bgfx::Attrib::Count == BX_COUNTOF(s_attribName));
+
             std::pair<unsigned int, const char*> GetVaryingLocationAndNewNameForName(const char* name)
             {
                 // For OpenGL and Metal platforms, we have an issue where we have a hard limit on the number shader attributes supported.
                 // To work around this issue, instead of mapping our attributes to the most similar bgfx::attribute, instead replace
                 // the first attribute encountered with the symbol bgfx uses for attribute 0 and increment for each subsequent attribute encountered.
                 // This will cause our shader to have nonsensical naming, but will allow us to efficiently "pack" the attributes.
-                UNUSED(name);
                 m_genericAttributesRunningCount++;
+                if (IsInstance(name))
+                {
+                    return {static_cast<unsigned int>(m_genericAttributesRunningCount - 1), s_attribInstanceName[m_instanceAttributeIndex++]};
+                }
                 if (m_genericAttributesRunningCount >= static_cast<unsigned int>(bgfx::Attrib::Count))
                     throw std::runtime_error("Cannot support more than 18 vertex attributes.");
 
                 return {static_cast<unsigned int>(m_genericAttributesRunningCount - 1), s_attribName[static_cast<unsigned int>(m_genericAttributesRunningCount - 1)]};
             }
+            unsigned int m_instanceAttributeIndex{0};
         };
 
+        class VertexVaryingInTraverserMetal final : private VertexVaryingInTraverser
+        {
+        public:
+            static void Traverse(TProgram& program, IdGenerator& ids, std::unordered_map<std::string, std::string>& replacementToOriginalName)
+            {
+                auto intermediate{program.getIntermediate(EShLangVertex)};
+                VertexVaryingInTraverserMetal traverser{};
+                intermediate->getTreeRoot()->traverse(&traverser);
+                VertexVaryingInTraverserMetal::Traverse(intermediate, ids, replacementToOriginalName, traverser);
+            }
+
+        private:
+            static void Traverse(TIntermediate* intermediate, IdGenerator& ids, std::unordered_map<std::string, std::string>& replacementToOriginalName, VertexVaryingInTraverser& traverser)
```

Comment:
I think it is easier if you made this an instance function instead of static since you are passing in the instance anyways?

Reply:
Done.